### ISSUE TITLE
How to get the loss and gradient for each data point given a lgb.dataset and lgb.booster object?

### ISSUE THREAD

**Issue by aBeginner666** [microsoft/LightGBM#4917]
Hi there, 

Given a lgb.dataset object and a lgb.booster object, how can I obtain the gradient and loss for each data point? Is it possible to get these info thru any API or are these info completely hidden from users? Maybe the loss is easy to calculate, just by predicting on the train set and feed into the loss function, but what about the gradient (particularly at any iteration, not just the last one)?

Any help is appreciated!

**Comment by shiyu1994** [comment]
@aBeginner666 Thanks for using LightGBM. Currently the gradients and loss of each data point is hidden from the users. With only a trained model, it is hard to recover the gradients per iteration. The gradients are not recorded in the model.

Is it acceptable for your case to get the gradients during training of the model?

**Comment by aBeginner666** [comment]
Hi Shiyu, 

Thanks for your response. I think it's acceptable to get the gradients during training of the model, and then I will try to save gradient along the way. Is it feasible? What should I do?

Thanks!

**Comment by nsarang** [comment]
@shiyu1994 Is it possible to get the gradients of each iteration during training?
